feat: Initial attempt at napi-rs

b-zee opened the pull request:
An initial attempt using napi-rs and using async/runtimes in a robust way.

One challenge I'm heading into is that the wrapped `Safe` instance can't really be borrowed for a Future to execute. For now I've cloned the instance, but I'm unsure of the side effects of cloning it, as I'm sure there's lots of configuration and internals involved.

b-zee committed:
feat: Initial attempt at napi-rs

happybeing commented:
Hey @b-zee I'm curious about this. I don't know anything about native modules but think MaidSafe were using `neon` for this and guess for some reason you're looking at switching to `napi-rs`. Can you explain a little for me?

b-zee commented:
Hey @happybeing! The reason is mostly in that `napi-rs` aligns more closely with the so-called N-API which is what Neon also builds on since the past few months. Except Neon abstracts a lot away which doesn't allow us to have the control we want. But, Neon is very much being developed, so anything I say here can be outdated already. (I've come across a migration guide they're [working on at this moment](https://github.com/neon-bindings/neon/blob/migration-experiment/MIGRATION_GUIDE.md) that also shows some of the changes and concepts.)

The reason I would want some more control is mainly about threading and async. Neon did make it easy to do some relatively easy asynchronous tasks with their (Task) API, but these are very narrow use-cases.

The Safe Rust API uses async heavily, and that means we need to use a runtime. Node.js is single-threaded and that means the runtime should not be run on the main thread, but on a separate thread. The challenge then becomes on how to 'send' data from the main thread to the runtime and assuring the Node.js garbage collector doesn't destroy anything in the meantime.

I've [made a topic on the Safe Dev Forum](https://forum.safedev.org/t/adventures-in-rust-node-js-and-safe/2959?u=bzee) alluding to the challenge and trying to explain a few concepts along the way.

b-zee committed:
feat: Keypair, async, tests

b-zee committed:
feat: TS integration, modular structure

b-zee committed:
feat: simplify conversion with serde and more

Defined a Jest environment for all tests to run in.
Tweaked TypeScript settings
Files function to test conversion

b-zee committed:
cleanup: simplified and extended some interfaces

b-zee committed:
style: cargo fmt

b-zee committed:
refactor: restructured modules

b-zee committed:
feat: JS types for NRS function and test

b-zee committed:
chore: removed superfluous test file

b-zee committed:
feat: add nrs get container fn

b-zee committed:
feat: handle Rust Enums in JS (d.ts/test)

b-zee committed:
feat: nrs add subname

b-zee committed:
chore: update sn_api

b-zee committed:
feat/style: add authd status and small refactor

b-zee committed:
chore: update cargo and add trace logs

b-zee committed:
feat: add authd_client methods

b-zee committed:
feat: SafeAuthdClient getter for struct member

b-zee committed:
chore: update dependencies and prepare release

b-zee committed:
chore: build on installing this package

b-zee committed:
feat: implement fetch

fetch returns an enum that is by default serialised by napi-rs in a way
that loses the variant information. This was solved by wrapping the enum
in a type that specifically will keep this information.

b-zee committed:
feat: CI attempt

b-zee committed:
chore: update sn_api dependency

b-zee committed:
chore: update dependencies; remove tokio compat

b-zee committed:
feat: update sn_container action

b-zee committed:
chore: update sn_api and Cargo.lock
